src/Sarif.Viewer.VisualStudio/Tags/SarifLocationTagger.cs

```diff
@@ -0,0 +1,573 @@
+        {
+            List<SarifLocationTag> tagsToDispose = new List<SarifLocationTag>();
+
+            using (TagListLock.EnterWriteLock())
+            {
+                // There is no need to re-add the tags from the run Id to SARIF tag dictionary as the
+                // objects exist in both dictionaries and only need to be disposed once.
+                tagsToDispose.AddRange(SourceCodeFileToSarifTags.Values.SelectMany(sarifTags => sarifTags));
+
+                SourceCodeFileToSarifTags.Clear();
+                RunIdToSarifTags.Clear();
+            }
+
+            tagsToDispose.ForEach(tagToDispose => tagToDispose.Dispose());
+
+            // Note that the lock wrapper implementation does not support IDispose
+            // and therefore we still must dispose the "inner lock" held by the wrapper.
+            TagListLock.InnerLock.Dispose();
+            SarifTaggersLock.InnerLock.Dispose();
+        }
+
+            /// <inheritdoc/>
+        public ISarifLocationTag AddTag(Region sourceRegion, TextSpan documentSpan, int runId, TextMarkerTag tag)
+        {
+            // Start an update so that even on a call (where the caller is not batching using Update themselves)
+            // that a "tags changed" event is fired to update VS's editor.
+            using (this.Update())
+            {
+                using (TagListLock.EnterWriteLock())
+                {
+                    SarifLocationTag existingSarifTag = this.sarifTags.FirstOrDefault(
+                        (sarifTag) =>
+                            sarifTag.SourceRegion.ValueEquals(sourceRegion));
+
+                    if (existingSarifTag != null)
+                    {
+                        return existingSarifTag;
+                    }
+
+                    IPersistentSpan persistentSpan = this.persistentSpanFactory.Create(
+                        this.textBuffer.CurrentSnapshot,
+                        startLine: documentSpan.iStartLine,
+                        startIndex: documentSpan.iStartIndex,
+                        endLine: documentSpan.iEndLine,
+                        endIndex: documentSpan.iEndIndex,
+                        trackingMode: SpanTrackingMode.EdgeInclusive);
+
+                    SarifLocationTag newSarifTag = new SarifLocationTag(
+                        persistentSpan,
+                        sourceRegion,
+                        runId,
+                        tag);
+
+                    this.sarifTags.Add(newSarifTag);
+                    newSarifTag.PropertyChanged += SarifTagPropertyChanged;
+
+                    if (RunIdToSarifTags.TryGetValue(runId, out List<SarifLocationTag> sarifTagsForRun))
+                    {
+                        sarifTagsForRun.Add(newSarifTag);
+                    }
+                    else
+                    {
+                        RunIdToSarifTags.Add(runId, new List<SarifLocationTag> { newSarifTag });
+                    }
```

Comment:
I like the pattern
```
if (!x.TryGetValue(key, out List<T> list)
{
    list = new List<T>();
    x[key] = list;
}

list.Add(thing);
```

Reply:
Yep, that works, but... it does require more list manipulation than the existing code since the a new list is initialized with the new object and hence no need to call "add" for a new list. So technically the existing code is more efficient (minorly).

---
In reply to: [482236299](https://github.com/microsoft/sarif-visualstudio-extension/pull/191#discussion_r482236299) [](ancestors = 482236299)